beacon-chain/verification/blob.go

```diff
@@ -0,0 +1,285 @@
+	if bv.fc.IsCanonical(bv.blob.ParentRoot()) {
+		return nil
+	}
+	return ErrSidecarDescendsFromFinalized
+}
+
+// SidecarInclusionProven represents the spec verification:
+// [REJECT] The sidecar's inclusion proof is valid as verified by verify_blob_sidecar_inclusion_proof(blob_sidecar).
+func (bv *BlobVerifier) SidecarInclusionProven() (err error) {
+	defer bv.recordResult(RequireSidecarInclusionProven, &err)
+	if err := blocks.VerifyKZGInclusionProof(bv.blob); err != nil {
+		log.WithError(err).WithFields(logging.BlobFields(bv.blob)).Debug("sidecar inclusion proof verification failed")
+		return ErrSidecarInclusionProven
+	}
+	return nil
+}
+
+// SidecarBlobCommitmentProven represents the spec verification:
+// [REJECT] The sidecar's blob is valid as verified by
+// verify_blob_kzg_proof(blob_sidecar.blob, blob_sidecar.kzg_commitment, blob_sidecar.kzg_proof).
+func (bv *BlobVerifier) SidecarBlobCommitmentProven() (err error) {
+	defer bv.recordResult(RequireSidecarBlobCommitmentProven, &err)
+	if err := bv.verifyBlobCommitment(bv.blob); err != nil {
+		log.WithError(err).WithFields(logging.BlobFields(bv.blob)).Debug("kzg commitment proof verification failed")
+		return ErrSidecarBlobCommitmentProven
+	}
+	return nil
+}
+
+// SidecarProposerExpected represents the spec verification:
+// [REJECT] The sidecar is proposed by the expected proposer_index for the block's slot
+// in the context of the current shuffling (defined by block_header.parent_root/block_header.slot).
+// If the proposer_index cannot immediately be verified against the expected shuffling, the sidecar MAY be queued
+// for later processing while proposers for the block's branch are calculated -- in such a case do not REJECT, instead IGNORE this message.
+func (bv *BlobVerifier) SidecarProposerExpected(ctx context.Context) (err error) {
+	defer bv.recordResult(RequireSidecarProposerExpected, &err)
+	idx, cached := bv.pc.Proposer(bv.blob.ParentRoot(), bv.blob.Slot())
+	if !cached {
+		pst, err := bv.parentState(ctx)
+		if err != nil {
+			log.WithError(err).WithFields(logging.BlobFields(bv.blob)).Debug("state replay to parent_root failed")
+			return ErrSidecarProposerExpected
+		}
+		idx, err = bv.pc.ComputeProposer(ctx, bv.blob.ParentRoot(), bv.blob.Slot(), pst)
+		if err != nil {
+			log.WithError(err).WithFields(logging.BlobFields(bv.blob)).Debug("error computing proposer index from parent state")
+			return ErrSidecarProposerExpected
+		}
+	}
+	if idx != bv.blob.ProposerIndex() {
+		log.WithError(ErrSidecarProposerExpected).
+			WithFields(logging.BlobFields(bv.blob)).WithField("expected_proposer", idx).
+			Debug("unexpected blob proposer")
+		return ErrSidecarProposerExpected
+	}
+	return nil
+}
+
+func (bv *BlobVerifier) parentState(ctx context.Context) (state.BeaconState, error) {
+	if bv.parent != nil {
+		return bv.parent, nil
+	}
+	st, err := bv.sr.StateByRoot(ctx, bv.blob.ParentRoot())
+	if err != nil {
+		return nil, err
+	}
+	bv.parent = st
```

Comment:
It's not clear to me how useful is this caching here, the parent state is called only for the valid proposer signature and proposer index computation only. Is there a way of sharing this among verifiers for all sidecars with the same block root?

Reply:
My thinking here is that if we don't have the signature cached, we may also not have the proposer cached, so we should keep the state until the verifier is done rather than regenerate it when there is a miss in both places.

Reply:
I'm also assuming other verifiers for the same root will typically run semi-sequentially so by the time the second one comes along the result of the first will be cached.